File: .github/scripts/analyze.py
import os
import subprocess
from google import genai

# 최신 SDK 클라이언트 설정
client = genai.Client(api_key=os.environ["GEMINI_API_KEY"])
MODEL_ID = "gemini-1.5-flash"

def get_target_files():
    """분석할 C++ 파일을 결정합니다."""
    # 1. 최근 커밋 변경 사항 확인
    try:
        output = subprocess.check_output(['git', 'diff', '--name-only', 'HEAD~1', 'HEAD']).decode('utf-8')
        files = [f for f in output.splitlines() if f.endswith(('.cc', '.cpp'))]
    except:
        files = []

    # 2. 변경 사항이 없으면 전체에서 C++ 파일 탐색
    if not files:
        try:
            output = subprocess.check_output(['git', 'ls-tree', '-r', 'HEAD', '--name-only']).decode('utf-8')
            files = [f for f in output.splitlines() if f.endswith(('.cc', '.cpp'))]
        except:
            files = []
            
    return files

def analyze_and_save():
    files = get_target_files()
    
    if not files:
        print("No C++ files found to analyze.")
        return

    for file_path in files:
        if not os.path.exists(file_path): continue
        
        print(f"Analyzing: {file_path}")
        with open(file_path, 'r', encoding='utf-8') as f:
            code = f.read()

        prompt = f"""
        당신은 알고리즘 전문가입니다. 다음 C++ 코드를 분석하여:
        1. 핵심 로직의 흐름도를 Mermaid.js flowchart TD 형식으로 작성하세요.
        2. 시간 복잡도와 공간 복잡도를 한국어로 짧게 설명하세요.
        
        결과는 마크다운 형식으로 작성해주시고, Mermaid 코드는 반드시 ```mermaid 블록으로 감싸주세요.
        코드:
        {code}
        """

        try:
            # 최신 SDK 호출 방식
            response = client.models.generate_content(
                model=MODEL_ID,
                contents=prompt
            )
            
            target_dir = os.path.dirname(file_path)
            analysis_path = os.path.join(target_dir, "ANALYSIS.md")
            
            with open(analysis_path, 'w', encoding='utf-8') as f:
                f.write(f"# 📊 Algorithm Analysis: {os.path.basename(file_path)}\n\n")
                f.write(response.text)
            print(f"Successfully created ANALYSIS.md in {target_dir}")
            
        except Exception as e:
            print(f"Error analyzing {file_path}: {e}")

if __name__ == "__main__":
    analyze_and_save()

File: .github/scripts/analyze.mjs
import fs from 'fs'
import path from 'path'
import fetch from 'node-fetch'

const files = process.argv[2]?.split('\n').filter(Boolean)
const API_KEY = process.env.GEMINI_API_KEY

if (!files || files.length === 0) {
  console.log('No C++ files changed.')
  process.exit(0)
}

const PROMPT_TEMPLATE = (code) => `
당신은 알고리즘 및 자료구조에 능숙한 전문가입니다.

아래에 제공된 C++ 코드는 알고리즘 문제의 풀이 코드입니다.
이 코드를 기반으로 다음 작업을 수행하세요.

1. 전체 알고리즘의 핵심 로직 흐름을 분석하고,
   이를 Mermaid.js의 flowchart TD 문법으로 시각화하세요.

2. 시간 복잡도와 공간 복잡도를
   Big-O 표기법과 함께 한국어로 간단히 설명하세요.

출력 규칙:
- Markdown 형식으로 작성하세요.
- Mermaid 다이어그램은 반드시 \`\`\`mermaid 블록을 사용하세요.
- 불필요한 서론은 작성하지 마세요.

[C++ 코드]
\`\`\`cpp
${code}
\`\`\`
`

async function generateAnalysis(filePath) {
  const code = fs.readFileSync(filePath, 'utf8')
  const dir = path.dirname(filePath)
  const outputPath = path.join(dir, 'ANALYSIS.md')

  const res = await fetch(
    'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=' +
      API_KEY,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [
          {
            parts: [{ text: PROMPT_TEMPLATE(code) }],
          },
        ],
      }),
    }
  )

  const json = await res.json()
  const text = json.candidates?.[0]?.content?.parts?.[0]?.text

  if (!text) {
    console.error(`Failed to generate analysis for ${filePath}`)
    return
  }

  fs.writeFileSync(outputPath, text, 'utf8')
  console.log(`Generated: ${outputPath}`)
}

for (const file of files) {
  await generateAnalysis(file)
}
